### DISCUSSION TITLE
How was the URDF constructed?

### DISCUSSION THREAD

**Discussion by one-for-all** [upkie/upkie#459] [Software] [answered]
Hi, I am wondering how did you create the URDF of this robot?

Did you just write it manually by adjusting the numbers until they look right? or used some software to assemble it visually, and exported from there?

What I usually do is assemble the robot in a CAD software such as Onshape, and export it into a URDF. However, it seems that your robot parts are stl/3mf files (i.e. mesh, not parametric modeling), and I think you mentioned that you designed them with Blender. In that case, I think it is rather difficult if not impossible to assemble the parts in a CAD software, which gave me hiccups.

E.g. I am trying to mate the Raspberry Pi and the head plate. But because the plate is mesh, it doesn't really have a circular hole for me to mate with:
![Screenshot 2024-11-13 at 9 03 11 PM](https://github.com/user-attachments/assets/79befeb3-418f-4656-9c12-6a90f70c1f79)

Thanks.

**Comment by stephane-caron** [comment, marked as answer]
Originally the URDF description was made using the visualization scripts that are now in [upkie_description](https://github.com/upkie/upkie_description/), such as [`display_joints.py`](https://github.com/upkie/upkie_description/blob/e20e970d75900bdbcbddfc97bd45cf2073145733/examples/display_joints.py) and [`display_frames.py`](https://github.com/upkie/upkie_description/blob/e20e970d75900bdbcbddfc97bd45cf2073145733/examples/display_frames.py), positioning frames manually until the corresponding meshes align. Basically moving things around until visually all parts met satisfactorily :sweat_smile: 

There is indeed a Blender project as well! It can be downloaded from the project page on Hackaday.io: [`upkie.blend`](https://cdn.hackaday.io/files/1857297946229536/upkie-blender.zip). It was not used to make the URDF; rather, I made it later (as a reference for other makers) when we assembled more robots and wrote the [build instructions](https://github.com/upkie/upkie/wiki).

Historically, the reason why many parts are only available as meshes is that they come from the [3D printing files of the mjbots quad A1](https://github.com/mjbots/quad/tree/4f7413238371d73bedaaff77f93c7d7ea1a6d8fd/hw), which we can build upon since it is open source. Newer parts in [upkie/parts](https://github.com/upkie/parts) were made with FreeCAD and we distribute the CAD files as well, but so far the previous parts are STL-only. PRs are welcome and some users have contributed redesigns with proper CAD software (like [this one](https://github.com/upkie/parts/pull/15) or [this one](https://github.com/upkie/parts/pull/14)).

**Reply by one-for-all** [reply]
Got it, and thanks for the description and links! Very useful.

I found a way to turn stl/3mf into editable step file for CAD, though that method might not be perfectly reliable.

I was working on assembling the robot in CAD when I posted the above question, and I am thinking of creating a step-by-step animation of the assembly procedure based on your build instructions.

I will post if I work out something useful. 🎉
